src/magentic/chat_model/openai_chat_model.py

```diff
@@ -336,52 +388,62 @@ def complete(
             seed=self.seed,
             stop=stop,
             temperature=self.temperature,
-            functions=openai_functions,
-            function_call=(
-                {"name": openai_functions[0]["name"]}
-                if len(openai_functions) == 1 and not allow_string_output
+            tools=[schema.to_dict() for schema in tool_schemas],
+            tool_choice=(
+                tool_schemas[0].as_tool_choice()
+                if len(tool_schemas) == 1 and not allow_string_output
                 else None
             ),
         )
 
-        # Azure OpenAI sends a chunk with empty choices first
         first_chunk = next(response)
+        # Azure OpenAI sends a chunk with empty choices first
         if len(first_chunk.choices) == 0:
             first_chunk = next(response)
+        if (
+            first_chunk.choices[0].delta.content is None
+            and first_chunk.choices[0].delta.tool_calls is None
+        ):
+            first_chunk = next(response)
+        response = chain([first_chunk], response)  # Replace chunk containing content
 
-        response = chain([first_chunk], response)  # Replace first chunk
+        if first_chunk.choices[0].delta.content is not None:
+            if not allow_string_output:
+                msg = (
+                    "String was returned by model but not expected. You may need to update"
+                    " your prompt to encourage the model to return a specific type."
+                )
+                raise ValueError(msg)
+            streamed_str = StreamedStr(
+                chunk.choices[0].delta.content
+                for chunk in response
+                if chunk.choices[0].delta.content is not None
+            )
+            if streamed_str_in_output_types:
+                return AssistantMessage(streamed_str)  # type: ignore[return-value]
+            return AssistantMessage(str(streamed_str))
 
-        function_schema = self._select_function_schema(first_chunk, function_schemas)
-        if function_schema:
+        if first_chunk.choices[0].delta.tool_calls is not None:
+            # TODO: if MultiFunctionCall in output_types ...
+            tool_schema = self._select_tool_schema(first_chunk, tool_schemas)
             try:
-                content = function_schema.parse_args(
-                    chunk.choices[0].delta.function_call.arguments
-                    for chunk in response
-                    if chunk.choices[0].delta.function_call
-                    and chunk.choices[0].delta.function_call.arguments is not None
+                # Take only the first tool_call, silently ignore extra chunks
+                content = tool_schema.parse_response(
+                    takewhile(
+                        lambda chunk: chunk.choices[0].delta.tool_calls[0].index == 0,  # type: ignore[index]
+                        response,
+                    )
                 )
```

Comment:
I'm guessing the intention will be to support the "parallel" function calling when this is complete, right?

Reply:
Yes, adding new `ParallelFunctionCall` and `AsyncParallelFunctionCall` that will allow you use that.

We use streaming responses to support `Iterable`, `AsyncIterable` and other return types that can start being used before the full response has been received. When we determine that one of these is being returned we can hand the stream off to the user so they can start using it without us blocking until the end. The issue with tool_calls is that we can't tell whether the model is calling one or multiple tools from the first few chunks, so in this case we will see if the user has typed their function as returning `ParallelFunctionCall` or `FunctionCall` (/other structured object) and parse based on that. Because the stream is being handed over to the user any check we put for whether there is a second tool_call will happen while they are consuming this stream. The options I can see to deal with this are 1) raise an exception when the stream gets to a second tool call or 2) end the stream after the first tool call. I opted for (2) since I don't want to raise exceptions for users after it seems like they have left magentic and are back in their own code. Potentially could log a warning that a second tool call was returned.